one/src/events.rs

```diff
@@ -48,17 +71,221 @@ async fn slurp(opts: SlurpOpts) -> Result<()> {
+    }
+    Ok(())
+}
+
+async fn validate(opts: ValidateOpts) -> Result<()> {
+    let pool = SqlitePool::from_store_dir(opts.store_dir).await.unwrap();
+    let block_store = SQLiteBlockStore::new(pool.clone()).await.unwrap();
+    let root_store = SQLiteRootStore::new(pool).await.unwrap();
+
+    // Validate that the CID is either DAG-CBOR or DAG-JOSE
+    if (opts.cid.codec() != 0x71) && (opts.cid.codec() != 0x85) {
+        return Err(anyhow!("CID {} is not a valid Ceramic event", opts.cid));
+    }
+
+    // If the CID is a DAG-JOSE, the event is either a signed Data Event or a signed Init Event, both of which can be
+    // validated similarly.
+    if opts.cid.codec() == 0x85 {
+        let controller = validate_data_event_envelope(&opts.cid, &block_store).await;
+        match controller {
+            Ok(controller) => {
+                println!(
+                    "DataEvent({}) validated as authored by Controller({})",
+                    opts.cid, controller
+                )
+            }
+            Err(e) => println!("{} failed with error: {:?}", opts.cid, e),
+        }
+        return Ok(());
+    }
+    // If the CID is a DAG-CBOR, the event is either a Time Event or an unsigned Init Event. In this case, we'll pull
+    // the block from the store and use the presence of the "proof" field to determine whether this is a Time Event or
+    // an unsigned Init Event.
+    //
+    // When validating events from Recon, the block store will be a CARFileBlockStore that wraps the CAR file received
+    // over Recon.
+    if let Some(block) = block_store.get(opts.cid).await? {
+        let event = CborValue::parse(&block)?;
+        if event.get_key("proof").is_some() {
+            let timestamp = validate_time_event(
+                &opts.cid,
+                &block_store,
+                &root_store,
+                &HttpEthRpc::new(opts.ethereum_rpc_url),
+            )
+            .await;
+            match timestamp {
+                Ok(timestamp) => {
+                    let rfc3339 = Utc
+                        .timestamp_opt(timestamp, 0)
+                        .unwrap()
+                        .to_rfc3339_opts(SecondsFormat::Secs, true);
+                    println!(
+                        "TimeEvent({}) validated at Timestamp({}) = {}",
+                        opts.cid, timestamp, rfc3339
+                    )
+                }
+                Err(e) => println!("{} failed with error: {:?}", opts.cid, e),
+            }
+        } else {
+            if event.get_key("data").is_some() {
+                println!("UnsignedDataEvent({}) is not signed", opts.cid);
+            }
+            validate_data_event_payload(&opts.cid, &block_store, None).await?;
+        }
     }
     Ok(())
 }
 
+async fn validate_data_event_envelope(cid: &Cid, block_store: &SQLiteBlockStore) -> Result<String> {
+    let block = block_store.get(cid.to_owned()).await?.unwrap();
```

Comment:
nitpick, I'd actually prefer to leave all of this out of this PR, and change this function to just panic with "not yet implemented", then keep the data event validation in a separate PR so we're not doing too many things at once in a single PR.  I'm mostly focusing my efforts in this review on the time event validation, not going to read this function very closely right now, but it would probably deserve it's own proper review at some point.